orangecontrib/prototypes/widgets/owinteractions.py

```diff
@@ -0,0 +1,242 @@
+"""
+Interactions widget
+"""
+from operator import attrgetter
+from itertools import chain
+
+import numpy as np
+
+from AnyQt.QtCore import Qt
+from AnyQt.QtGui import QStandardItem
+
+from Orange.data import Table, Domain, ContinuousVariable, StringVariable
+from Orange.widgets import gui
+from Orange.widgets.utils.itemmodels import DomainModel
+from Orange.widgets.utils.signals import Input, Output
+from Orange.widgets.utils.widgetpreview import WidgetPreview
+from Orange.widgets.widget import OWWidget, AttributeList, Msg
+from Orange.preprocess import Discretize, Remove
+from Orange.preprocess.discretize import EqualFreq
+import Orange.widgets.data.owcorrelations
+
+
+SIZE_LIMIT = 1000000
+
+
+class Interaction:
+	def __init__(self, disc_data):
+		self.data = disc_data
+		self.n_attrs = len(self.data.domain.attributes)
+		self.class_h = self.entropy(self.data.Y)
+		self.attr_h = np.zeros(self.n_attrs)
+		self.gains = np.zeros(self.n_attrs)
+		self.removed_h = np.zeros((self.n_attrs, self.n_attrs))
+		self.compute_gains()
+
+	@staticmethod
+	def distribution(ar):
+		nans = np.isnan(ar)
+		if nans.any():
+			if len(ar.shape) == 1:
+				ar = ar[~nans]
+			else:
+				ar = ar[~nans.any(axis=1)]
+		_, counts = np.unique(ar, return_counts=True, axis=0)
+		return counts / len(ar)
+
+	def entropy(self, ar):
+		p = self.distribution(ar)
+		return -np.sum(p * np.log2(p))
+
+	def compute_gains(self):
+		for attr in range(self.n_attrs):
+			self.attr_h[attr] = self.entropy(self.data.X[:, attr])
+			self.gains[attr] = self.attr_h[attr] + self.class_h \
+				- self.entropy(np.c_[self.data.X[:, attr], self.data.Y])
+
+	def __call__(self, attr1, attr2):
+		attrs = np.c_[self.data.X[:, attr1], self.data.X[:, attr2]]
+		self.removed_h[attr1, attr2] = self.entropy(attrs) + self.class_h - self.entropy(np.c_[attrs, self.data.Y])
+		return self.removed_h[attr1, attr2] - self.gains[attr1] - self.gains[attr2]
+
+
+class Heuristic:
+	def __init__(self, weights):
+		self.n = len(weights)
+		self.w = weights.copy()
+		self.a = np.arange(self.n)
```

Comment:
I'd prefer more descriptive variable names.

Reply:
Fixed in the latest commit